src/psyclone/tests/extract_test.py

```diff
@@ -0,0 +1,828 @@
+    schedule, _ = otrans.apply(schedule.children[1])
+    loop = schedule.children[1].children[0]
+    # Try extracting the Loop inside the OMP Parallel DO region
+    with pytest.raises(TransformationError) as excinfo:
+        _, _ = etrans.apply(loop)
+    assert ("Extraction of a Loop without its parent Directive is not "
+            "allowed.") in str(excinfo)
+
+
+def test_loop_no_directive_gocean1p0():
+    ''' Test that applying ExtractRegionTrans on a Loop without its
+    parent Directive when optimisations are applied in GOcean1.0 API
+    raises a TransformationError. '''
+    from psyclone.transformations import GOceanOMPParallelLoopTrans
+
+    etrans = ExtractRegionTrans()
+    otrans = GOceanOMPParallelLoopTrans()
+
+    # Test a Loop nested within the OMP Parallel DO Directive
+    _, invoke_info = parse(os.path.join(GOCEAN_BASE_PATH,
+                                        "single_invoke_three_kernels.f90"),
+                           api=GOCEAN_API)
+    psy = PSyFactory(GOCEAN_API, distributed_memory=False).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    # Apply GOceanOMPParallelLoopTrans to the first Loop
+    schedule, _ = otrans.apply(schedule.children[0])
+    loop = schedule.children[0].children[0]
+    # Try extracting the Loop inside the OMP Parallel DO region
+    with pytest.raises(TransformationError) as excinfo:
+        _, _ = etrans.apply(loop)
+    assert ("Extraction of a Loop without its parent Directive is not "
+            "allowed.") in str(excinfo)
+
+
+def test_no_parent_ompdirective():
+    ''' Test that applying ExtractRegionTrans on an orphaned
+    OMPDoDirective without its ancestor OMPRegionDirective when
+    optimisations are applied raises a TransformationError. '''
+    from psyclone.transformations import OMPParallelTrans, \
+        Dynamo0p3OMPLoopTrans
+
+    etrans = ExtractRegionTrans()
+    ltrans = Dynamo0p3OMPLoopTrans()
+    otrans = OMPParallelTrans()
+
+    _, invoke_info = parse(os.path.join(DYNAMO_BASE_PATH,
+                                        "4.13_multikernel_invokes_w3.f90"),
+                           api=DYNAMO_API)
+    psy = PSyFactory(DYNAMO_API, distributed_memory=False).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+
+    # Apply the OpenMP Loop transformation to every loop in the Schedule
+    for child in schedule.children:
+        if isinstance(child, Loop):
+            newschedule, _ = ltrans.apply(child)
+    # Enclose all of these loops within a single OpenMP Parallel region
+    schedule, _ = otrans.apply(newschedule.children)
+
+    orphaned_directive = schedule.children[0].children[1]
+    with pytest.raises(TransformationError) as excinfo:
+        _, _ = etrans.apply(orphaned_directive)
+    assert ("Extraction of an orphaned Directive without its ancestor "
+            "Directive is not allowed.") in str(excinfo)
+
+
+def test_no_parent_accdirective():
```

Comment:
They test different sub-classes of orphaned directives (OMP and ACC) so I left them as they are.

Reply:
On the face of it, yes, but in fact they both exercise the same check inside `ExtractRegionTrans._validate()`. Our approach is to write tests to acheive 100% coverage rather than to check every possible combination since that rapidly becomes impossible.

Reply:
After modifying the constraint for orphaned directives in `ExtraxctRegionTrans` I realised your reasoning is correct. I deleted the preceding (OMP) test to so there is a better balance between Dynamo0.3 and GOcean1.0 code.